File: Cindy/Signal/FrameObservation.swift
import Foundation
import CoreGraphics
import ImageIO

/// Face bounding box in Vision's normalised coordinates (origin bottom-left, 0…1).
struct FaceObservation: Equatable, Sendable {
    var boundingBox: CGRect
    var confidence: Float

    var area: Float { Float(boundingBox.width * boundingBox.height) }
    var centerY: Float { Float(boundingBox.midY) }
    var centerX: Float { Float(boundingBox.midX) }
}

enum PoseJoint: String, CaseIterable, Sendable {
    case nose, leftShoulder, rightShoulder, leftHip, rightHip
}

struct PosePoint: Equatable, Sendable {
    var x: Float
    var y: Float
    var confidence: Float
}

/// Selected body-pose landmarks in normalised coordinates (origin bottom-left).
struct BodyPoseObservation: Equatable, Sendable {
    var joints: [PoseJoint: PosePoint]

    subscript(joint: PoseJoint) -> PosePoint? { joints[joint] }

    /// Mean y of the given joints, weighted only by presence; confidence = min of the used joints.
    func meanY(of jointsToUse: [PoseJoint]) -> (y: Float, confidence: Float)? {
        let points = jointsToUse.compactMap { joints[$0] }
        guard !points.isEmpty else { return nil }
        let y = points.map(\.y).reduce(0, +) / Float(points.count)
        let confidence = points.map(\.confidence).min() ?? 0
        return (y, confidence)
    }

    var noseY: Float? { joints[.nose]?.y }
    var shoulderY: Float? { meanY(of: [.leftShoulder, .rightShoulder])?.y }
    var hipY: Float? { meanY(of: [.leftHip, .rightHip])?.y }
    var overallConfidence: Float { joints.values.map(\.confidence).max() ?? 0 }
}

/// Everything the vision stage extracted from one camera frame.
struct FrameObservation: Equatable, Sendable {
    /// Presentation timestamp of the frame in seconds (monotonic).
    var timestamp: TimeInterval
    var face: FaceObservation?
    var pose: BodyPoseObservation?
    /// Orientation that produced the face hit (nil when no face was found).
    var orientation: CGImagePropertyOrientation?

    init(timestamp: TimeInterval, face: FaceObservation? = nil, pose: BodyPoseObservation? = nil,
         orientation: CGImagePropertyOrientation? = nil) {
        self.timestamp = timestamp
        self.face = face
        self.pose = pose
        self.orientation = orientation
    }
}

File: android/core/src/main/kotlin/me/raddatz/cindy/core/signal/FrameObservation.kt
package me.raddatz.cindy.core.signal

import me.raddatz.cindy.core.Rect
import kotlin.math.abs
import kotlin.math.hypot

/**
 * Face bounding box in normalised coordinates (origin bottom-left, 0…1, as Vision reports it).
 * The Android detector has to flip its top-left y before building one.
 */
data class FaceObservation(val boundingBox: Rect, val confidence: Float) {
    /** Uses the standardised (absolute) size, like `CGRect.width`/`height`. */
    val area: Float get() = (abs(boundingBox.width) * abs(boundingBox.height)).toFloat()
    val centerY: Float get() = boundingBox.midY.toFloat()
    val centerX: Float get() = boundingBox.midX.toFloat()
}

enum class PoseJoint(val rawValue: String) {
    NOSE("nose"),
    LEFT_SHOULDER("leftShoulder"),
    RIGHT_SHOULDER("rightShoulder"),
    LEFT_HIP("leftHip"),
    RIGHT_HIP("rightHip"),
}

data class PosePoint(val x: Float, val y: Float, val confidence: Float)

/** A measured scalar with the confidence it carries (Swift `(value:, confidence:)` tuples). */
data class MeasuredValue(val value: Float, val confidence: Float)

/** Selected body-pose landmarks in normalised coordinates (origin bottom-left). */
data class BodyPoseObservation(
    val joints: Map<PoseJoint, PosePoint>,
    /**
     * Width and height of the oriented image relative to its longer side, so distances between
     * joints have the same unit on both axes and in every orientation.
     */
    val xScale: Float = 1f,
    val yScale: Float = 1f,
) {
    operator fun get(joint: PoseJoint): PosePoint? = joints[joint]

    /** Mean y of the given joints, weighted only by presence; confidence = min of the used joints. */
    fun meanY(jointsToUse: List<PoseJoint>): MeasuredValue? {
        val points = jointsToUse.mapNotNull { joints[it] }
        if (points.isEmpty()) return null
        var sum = 0f
        for (point in points) sum += point.y
        val confidence = points.minOf { it.confidence }
        return MeasuredValue(sum / points.size.toFloat(), confidence)
    }

    val noseY: Float? get() = joints[PoseJoint.NOSE]?.y

    /** Distance between the shoulders in normalised image units; grows as the athlete gets closer. */
    val shoulderWidth: Float? get() = shoulderWidthSample?.value

    /** Shoulder width with the confidence of the weaker shoulder. */
    val shoulderWidthSample: MeasuredValue?
        get() {
            val left = joints[PoseJoint.LEFT_SHOULDER] ?: return null
            val right = joints[PoseJoint.RIGHT_SHOULDER] ?: return null
            return MeasuredValue(
                hypot((left.x - right.x) * xScale, (left.y - right.y) * yScale),
                minOf(left.confidence, right.confidence),
            )
        }

    val shoulderY: Float? get() = meanY(listOf(PoseJoint.LEFT_SHOULDER, PoseJoint.RIGHT_SHOULDER))?.value
    val hipY: Float? get() = meanY(listOf(PoseJoint.LEFT_HIP, PoseJoint.RIGHT_HIP))?.value
    val overallConfidence: Float get() = joints.values.maxOfOrNull { it.confidence } ?: 0f
}

/**
 * Image brightness: the squat signal (a closer body covers more of the bright ceiling),
 * measured when the brightness source runs and always in the debug recorder.
 */
data class FrameMetrics(
    /** Mean luma of the whole frame (0…1). */
    val lumaMean: Float? = null,
    /** Mean luma of the central half (width and height) of the frame. */
    val lumaCenter: Float? = null,
)

/** Everything the vision stage extracted from one camera frame. */
data class FrameObservation(
    /** Presentation timestamp of the frame in seconds (monotonic). */
    val timestamp: Double,
    val face: FaceObservation? = null,
    val pose: BodyPoseObservation? = null,
    /**
     * EXIF orientation (1…8, `CGImagePropertyOrientation` raw value on iOS) that produced the
     * face hit; null when no face was found. Only logged, never used by the detection logic.
     */
    val orientation: Int? = null,
    /** Same for the body-pose hit. */
    val poseOrientation: Int? = null,
    val metrics: FrameMetrics? = null,
)
